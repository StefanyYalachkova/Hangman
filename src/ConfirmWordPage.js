import { Button, TextField } from '@mui/material';

const ConfirmWordPage = (props) => {
    const MIN_LENGTH = 3;

    const { word, setWord, isValidInput } = props;
    const shouldDisableWordButton = word.length < MIN_LENGTH;

    const handleHangmanBoard = (event) => {
        event.preventDefault();
        props.onClick();
    };

    return (
        <form autoComplete="off">
            <span id="wordField">Word:</span>
            <TextField
                autoFocus={true}
                type="password"
                inputMode="tel"
                name="word"
                placeholder='Allow only alphabets'
                value={word}
                onChange={setWord}
                helperText={isValidInput
                    ? ''
                    : <span className="errorMessage">Need at least 3 letters.</span>
                }
            />
            <Button
                disabled={shouldDisableWordButton}
                name="wordButton" variant="contained"
                type="submit" onClick={handleHangmanBoard}
            > Confirm word
            </Button>
        </form>
    );
};

export { ConfirmWordPage };
