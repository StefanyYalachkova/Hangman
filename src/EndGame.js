import { Button } from '@mui/material';
import React, { useEffect } from 'react';

const EndGame = (props) => {
    const { word, checkWin, renderNewRound, setStartGamePage } = props;
    let message = '';
    let gameOver = false;

    const handleNewGame = (event) => {
        event.preventDefault();
        renderNewRound();
    };

    useEffect(() => {
        if (checkWin !== '') {
            setStartGamePage(true);
        };
    }, [handleNewGame]);

    if (checkWin === 'win') {
        message = 'Game over! You win!';
        gameOver = true;
    } else if (checkWin === 'loss') {
        message = `Game over! You lost! The word was: ${word}`;
        gameOver = true;
    };

    return (
        <>
            {gameOver
                ?
                <div className="winner-container">
                    <div className="winner">
                        <h2>{message}</h2>
                        <Button
                            autoFocus={true}
                            name="nemGameButton"
                            variant="contained"
                            type="submit"
                            onClick={handleNewGame}
                        > New Game
                        </Button>
                    </div>
                </div>
                : ''
            }
        </>
    );
};

export { EndGame };
