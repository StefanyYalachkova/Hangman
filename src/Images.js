import React from "react";
import image0 from './images/0.jpg';
import image1 from './images/1.jpg';
import image2 from './images/2.jpg';
import image3 from './images/3.jpg';
import image4 from './images/4.jpg';
import image5 from './images/5.jpg';
import image6 from './images/6.jpg';

const Images = (props) => {
    const { mistakes } = props;
    let mistake = mistakes.length;
    let image = '';

    if (mistake === 0) {
        image = <img src={image0} alt="image0" className="img" />
    } else if (mistake === 1) {
        image = <img src={image1} alt="image1" className="img" />
    } else if (mistake === 2) {
        image = <img src={image2} alt="image2" className="img" />
    } else if (mistake === 3) {
        image = <img src={image3} alt="image3" className="img" />
    } else if (mistake === 4) {
        image = <img src={image4} alt="image4" className="img" />
    } else if (mistake === 5) {
        image = <img src={image5} alt="image5" className="img" />
    } else if (mistake === 6) {
        image = <img src={image6} alt="image6" className="img" />
    };

    return image;
};

export { Images };
