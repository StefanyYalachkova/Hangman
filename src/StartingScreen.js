import { Button } from '@mui/material';
import { useEffect } from 'react';

const StartingScreen = (props) => {

    const handleNewGame = (event) => {
        event.preventDefault();
        props.onClick();
    };

    return (
        <div className="container" >
            <h1> H a n g m a n </h1>
            <Button
                autoFocus={true}
                name="nemGameButton"
                variant="contained"
                type="submit"
                onClick={handleNewGame}
            > New Game
            </Button>
        </div>
    );
};

export default StartingScreen;
