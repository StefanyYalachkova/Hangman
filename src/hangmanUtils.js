const isValidWord = (name, value) => {
    const isValidWordLength = value.length < 1 || value.length >= 3;

    if (name === 'word') {
        return isValidWordLength;
    };
};

const checkWin = (word, correctLetters, wrongLetters) => {
    let result = 'win';

    word.split('').forEach(letter => {
        if (!correctLetters.includes(letter)) {
            result = '';
        };
    });

    if (wrongLetters.length === 6) {
        result = 'loss';
    };

    return result;
};

export { isValidWord, checkWin };
