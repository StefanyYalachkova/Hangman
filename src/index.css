body {
  padding: 150px;
  background-color: #f1d2e0e1;
}

button[name="nemGameButton"] {
  font-family: sans-serif;
  background-color: rgb(0, 0, 255);
  font-weight: bold;
  padding: 20px;
  border-radius: 8px;
  height: 100%;
  cursor: pointer;
  margin: auto;
  position: relative;
  left: 45%;
}

button[name="wordButton"] {
  font-family: sans-serif;
  background-color: rgb(0, 0, 255);
  font-weight: bold;
  padding: 15px;
  border-radius: 8px;
  height: 100%;
  cursor: pointer;
  margin: auto;
  position: relative;
  left: 2%;
}

form,
#container,
h1 {
  text-align: center;
  padding: 50px;
}

.letter {
  text-align: inline-block;
  padding: 20px;
  font-weight: bold;
  font-size: x-large;
}

#wordField {
  padding: 20px;
  font-weight: bold;
  font-size: 20px;
}

h2 {
  text-align: inherit;
}

.errorMessage {
  color: red;
  font-size: 15px;
}

.img {
  display: block;
  margin-left: auto;
  margin-right: auto;
}

.wrongLetters {
  position: absolute;
  top: 300px;
  right: 500px;
  display: flex;
  flex-direction: column;
  text-align: right;
}

.word {
  display: flex;
  position: absolute;
  bottom: 150px;
  left: 42%;
}

.winner-container {
  background-color: rgba(109, 87, 87, 0.3);
  position: fixed;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  align-items: center;
  justify-content: center;
}

.winner {
  background: #9ac0e5;
  border-radius: 5px;
  box-shadow: 0 15px 10px 3px rgba(0, 0, 0, 0.1);
  padding: 20px;
  text-align: center;
}

.winner button {
  font-family: sans-serif;
  background-color: rgb(0, 0, 255);
  font-weight: bold;
  padding: 15px;
  border-radius: 8px;
  height: 100%;
  cursor: pointer;
  margin: auto;
  position: relative;
  left: 0%;
}