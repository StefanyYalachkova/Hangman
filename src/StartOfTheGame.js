import { useEffect, useRef } from 'react';
import { EndGame } from './EndGame';
import { Images } from './Images';
import { Word } from './Word';
import { WrongLetters } from './WrongLetters';

const StartOfTheGame = (props) => {
    const { wrongLetters, word, correctLetters, restartGame } = props;
    const { setStartGamePage, setWrongLetters, setCorrectLetters, renderNewRound, handleKeydown } = props;

    const boardRef = useRef(null)

    useEffect(() => {
        boardRef && boardRef.current && boardRef.current.focus();
    }, []);

    return (
        <div ref={boardRef}  onKeyDown={(event) => handleKeydown(event)} tabIndex={0}> 
            <Images
                wrongLetters={wrongLetters}
            />
            <Word
                word={word}
                correctLetters={correctLetters}
                restartGame={restartGame}
                setCorrectLetters={setCorrectLetters}
            />
            <WrongLetters
                wrongLetters={wrongLetters}
                restartGame={restartGame}
                setWrongLetters={setWrongLetters}
            />
            <EndGame
                word={word}
                correctLetters={correctLetters}
                wrongLetters={wrongLetters}
                setStartGamePage={setStartGamePage}
                renderNewRound={renderNewRound}
            />
        </div>
    );
};

export { StartOfTheGame };