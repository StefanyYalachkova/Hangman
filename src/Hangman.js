import { useState } from 'react';
import { ConfirmWordPage } from './ConfirmWordPage';
import { isValidWord } from './hangmanUtils';
import StartingScreen from './StartingScreen';
import { StartOfTheGame } from './StartOfTheGame';

const Hangman = () => {
    const [word, setWord] = useState('');
    const [viewConfirmWordPage, setViewConfirmWordPage] = useState(false);
    const [viewStartGamePage, setStartGamePage] = useState(false);

    const onWordPage = (event) => {
        const { name, value } = event.target;
        const regEx = /^[a-z]+$/;

        if (regEx.test(value)) {
            setWord(value);
        } else {
            setWord('');
        };
    };

    const handleConfirmWordPage = () => {
        setViewConfirmWordPage(true);
    };

    const handleStartGamePage = () => {
        setStartGamePage(true);
    };

    const renderStartGamePage = () => {
        return (
            <div>
                {viewStartGamePage
                    ? <StartOfTheGame
                        word={word}
                    />
                    : ''
                }
            </div>
        );
    };

    const renderConfirmWordPage = () => {
        return (
            <div>
                {viewConfirmWordPage
                    ? <ConfirmWordPage
                        setWord={onWordPage}
                        word={word}
                        isValidInput={isValidWord('word', word)}
                        onClick={handleStartGamePage}
                    />
                    : <StartingScreen
                        onClick={handleConfirmWordPage}
                    />
                }
            </div>
        );
    };

    return (
        <div>
            {viewStartGamePage
                ? renderStartGamePage()
                : renderConfirmWordPage()
            }
        </div>
    );
};

export { Hangman };
