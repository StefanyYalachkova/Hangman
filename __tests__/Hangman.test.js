import React from "react";
import renderer from 'react-test-renderer';
import { ConfirmWordPage } from "../src/ConfirmWordPage";
import { Hangman } from "../src/Hangman";
import { StartingScreen } from "../src/StartingScreen";
import { StartOfTheGame } from "../src/StartOfTheGame";

describe('Hangman test', () => {
    it('renders correctly', () => {
        const tree = renderer
            .create(<Hangman />);
        expect(tree).toMatchSnapshot();
    });
    it('renders correctly StartingScreen', () => {
        const handleConfirmWordPage = jest.fn();
        const restartGame = jest.fn();
        const tree = renderer
            .create(
                <StartingScreen
                    onClick={handleConfirmWordPage}
                    restartGame={restartGame}
                />);
        expect(tree).toMatchSnapshot();
    });
    it('renders correctly ConfirmWordPage', () => {
        const word = 'stef';
        const onWordPage = jest.fn();
        const handleStartGamePage = jest.fn();
        const isValidWord = jest.fn();

        const tree = renderer
            .create(
                <ConfirmWordPage
                    setWord={onWordPage}
                    word={word}
                    isValidInput={isValidWord('word', word)}
                    onClick={handleStartGamePage}
                />);
        expect(tree).toMatchSnapshot();
    });
    it('renders correctly StartOfTheGame', () => {
        const word = '';
        const correctLetters = [];
        const wrongLetters = [];
        const renderNewRound = jest.fn();
        const setStartGamePage = jest.fn();
        const restartGame = jest.fn();
        const setCorrectLetters = jest.fn();
        const setWrongLetters = jest.fn();

        const tree = renderer
            .create(
                <StartOfTheGame
                    wrongLetters={wrongLetters}
                    word={word}
                    correctLetters={correctLetters}
                    setStartGamePage={setStartGamePage}
                    restartGame={restartGame}
                    setCorrectLetters={setCorrectLetters}
                    setWrongLetters={setWrongLetters}
                    renderNewRound={renderNewRound}
                />);
        expect(tree).toMatchSnapshot();
    });
});
