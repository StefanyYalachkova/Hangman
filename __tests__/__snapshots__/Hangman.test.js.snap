// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hangman test renders correctly 1`] = `
<div>
  <div>
    <div
      className="container"
    >
      <h1>
         H a n g m a n 
      </h1>
      <button
        autoFocus={true}
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
        disabled={false}
        name="nemGameButton"
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="submit"
      >
         New Game
      </button>
    </div>
  </div>
</div>
`;

exports[`Hangman test renders correctly ConfirmWordPage 1`] = `
<form
  autoComplete="off"
>
  <span
    id="wordField"
  >
    Word:
  </span>
  <div
    className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
    inputMode="tel"
  >
    <div
      className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
      onClick={[Function]}
    >
      <input
        aria-describedby=":r0:-helper-text"
        aria-invalid={false}
        autoFocus={true}
        className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
        disabled={false}
        id=":r0:"
        name="word"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Allow only alphabets"
        required={false}
        type="password"
        value="stef"
      />
      <fieldset
        aria-hidden={true}
        className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="css-ihdtdm"
        >
          <span
            className="notranslate"
          >
            â€‹
          </span>
        </legend>
      </fieldset>
    </div>
    <p
      className="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained MuiFormHelperText-filled css-1wc848c-MuiFormHelperText-root"
      id=":r0:-helper-text"
    >
      <span
        className="errorMessage"
      >
        Need at least 3 letters.
      </span>
    </p>
  </div>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    name="wordButton"
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="submit"
  >
     Confirm word
  </button>
</form>
`;

exports[`Hangman test renders correctly StartOfTheGame 1`] = `
<div>
  <img
    alt="image0"
    className="img"
    src="test-file-stub"
  />
  <div
    className="word"
  />
  <div
    className="wrongLetters"
  >
    <div />
  </div>
  <div
    className="winner-container"
  >
    <div
      className="winner"
    >
      <h2>
        Game over! You win!
      </h2>
      <button
        autoFocus={true}
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
        disabled={false}
        name="nemGameButton"
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="submit"
      >
         New Game
        <span
          className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`Hangman test renders correctly StartingScreen 1`] = `
<div
  className="container"
>
  <h1>
     H a n g m a n 
  </h1>
  <button
    autoFocus={true}
    className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    name="nemGameButton"
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="submit"
  >
     New Game
    <span
      className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</div>
`;
